pkg/graph/schema/conversion_cel_type.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2025 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import (
+	"fmt"
+
+	"github.com/google/cel-go/cel"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiservercel "k8s.io/apiserver/pkg/cel"
+	"k8s.io/utils/ptr"
+)
+
+// GenerateSchemaFromCELTypes generates a JSONSchemaProps from a map of CEL types.
+// The provider is used to recursively introspect struct types and extract all fields.
+func GenerateSchemaFromCELTypes(typeMap map[string]*cel.Type, provider *apiservercel.DeclTypeProvider) (*extv1.JSONSchemaProps, error) {
+	fieldDescriptors := make([]fieldDescriptor, 0, len(typeMap))
+
+	for path, celType := range typeMap {
+		exprSchema, err := inferSchemaFromCELType(celType, provider)
+		if err != nil {
+			return nil, fmt.Errorf("failed to infer schema from type at %v: %w", path, err)
+		}
+		fieldDescriptors = append(fieldDescriptors, fieldDescriptor{
+			Path:   path,
+			Schema: exprSchema,
+		})
+	}
+
+	return generateJSONSchemaFromFieldDescriptors(fieldDescriptors)
+}
+
+// primitiveTypeToSchema converts a CEL primitive type name to a JSONSchemaProps.
+func primitiveTypeToSchema(typeName string) (*extv1.JSONSchemaProps, bool) {
+	switch typeName {
+	case "bool":
+		return &extv1.JSONSchemaProps{Type: "boolean"}, true
+	case "int", "uint":
+		return &extv1.JSONSchemaProps{Type: "integer"}, true
+	case "double":
+		return &extv1.JSONSchemaProps{Type: "number"}, true
+	case "string":
+		return &extv1.JSONSchemaProps{Type: "string"}, true
+	case "bytes":
+		return &extv1.JSONSchemaProps{Type: "string", Format: "byte"}, true
+	case "null_type":
```

Comment:
where does this null_type come from? my assumption is that null_type in json schema is something different. Also I believe that XPreserveUnknownFields can only apply to objects, whereas null type would also allow primitives